control/passivity.py

```diff
@@ -45,43 +84,126 @@ def ispassive(sys):
+    )
 
-    matrix_list = []
-    state_space_size = sys.nstates
-    for i in range(0, state_space_size):
-        for j in range(0, state_space_size):
-            if j <= i:
-                P = np.zeros_like(A)
-                P[i, j] = 1.0
-                P[j, i] = 1.0
-                matrix_list.append(make_LMI_matrix(P).flatten())
+    if rho is not None:
+        sys_matrix_list = _make_P_basis_matrices(
+            n, eps, 1.0, make_LMI_matrix)
+        sys_constants = -make_LMI_matrix(np.zeros_like(A), rho, eps, 1.0)
+    else:
+        sys_matrix_list = _make_P_basis_matrices(
+            n, 1.0, eps, make_LMI_matrix)
+        sys_constants = -make_LMI_matrix(np.zeros_like(A), eps, nu, 1.0)
 
-    coefficents = np.vstack(matrix_list).T
+    sys_coefficents = np.vstack(sys_matrix_list).T
 
-    constants = -np.vstack((
-        np.hstack((np.zeros_like(A),  - C.T)),
-        np.hstack((- C, -D - D.T)))
-    )
+    # LMI to ensure P is positive definite
+    P_matrix_list = _P_pos_def_constraint(n)
+    P_coefficents = np.vstack(P_matrix_list).T
+    P_constants = np.zeros((n, n))
 
+    # cost function
     number_of_opt_vars = int(
-        (state_space_size**2-state_space_size)/2 + state_space_size)
+        (n**2-n)/2 + n)
     c = cvx.matrix(0.0, (number_of_opt_vars, 1))
+    c = cvx.matrix(np.append(np.array(c), -1.0))
+
+    Gs = [cvx.matrix(sys_coefficents)] + [cvx.matrix(P_coefficents)]
+    hs = [cvx.matrix(sys_constants)] + [cvx.matrix(P_constants)]
 
     # crunch feasibility solution
     cvx.solvers.options['show_progress'] = False
-    sol = cvx.solvers.sdp(c,
-                          Gs=[cvx.matrix(coefficents)],
-                          hs=[cvx.matrix(constants)])
+    sol = cvx.solvers.sdp(c, Gs=Gs, hs=hs)
+    return sol["x"]
+
 
-    return (sol["x"] is not None)
+def get_output_fb_index(sys):
+    '''
+    Returns the output feedback passivity index for the input system. This is largest gain that can be placed in
+    positive feedback with a system such that the new interconnected system is passive.
+    Parameters
+    ----------
+    sys: An LTI system
+        System to be checked.
+
+    Returns
+    -------
+    float: 
+        The OFP index 
+    '''
+    sol = _solve_passivity_LMI(sys, nu=eps)
+    if sol is None:
+        return -np.inf
+    else:
+        return sol[-1]
+
+
+def get_input_ff_index(sys):
+    '''
+    Returns the input feedforward passivity (IFP) index for the input system. This is the largest gain that can be 
+    placed in negative parallel interconnection with a system such that the new interconnected system is passive.
+    Parameters
+    ----------
+    sys: An LTI system
+        System to be checked.
+
+    Returns
+    -------
+    float: 
+        The IFP index 
+    '''
+
+    sol = _solve_passivity_LMI(sys, rho=eps)
+    if sol is None:
+        return -np.inf
+    else:
+        return sol[-1]
+
+
+def ispassive(sys):
```

Comment:
But what about the cases, when you want to determine (boolean)
- input passivity (`nu != 0, rho = 0`), 
- output passivity (`nu = 0, rho != 0`)

now? One would have to use the `get_*_index()` functions. There is also no function to determine the relative passivity index.

Sorry if I created confusion earlier. I meant to make the function return values intuitive, not to remove functionality.

Reply:
You could determine input passivity with  `is_input_passive = get_input_ff_index(sys) > 0` (same for output passivity). I could add two more functions that do just that. 

Right now, I'm intentionally not fiddling with computing the relative passivity index, I need to go read the source that the Matlab documentation is citing "Sector Bounds in Stability Analysis and Control Design". If you think it should be included in this pull request, I'm totally willing to add it.

Reply:
> I could add two more functions that do just that.

Despite earlier discussions, I would vote for including that functionality into `ispassive(sys, nu, rho)`

Reply:
> Right now, I'm intentionally not fiddling with computing the relative passivity index, I need to go read the source that the Matlab documentation is citing "Sector Bounds in Stability Analysis and Control Design". If you think it should be included in this pull request, I'm totally willing to add it.

No that is fine. This can be defered to later PRs. Just raise a `NotImplemented` Exception instead of stating "This input is invalid". Same for I/O passivity index, I guess.

Reply:
> But what about the cases, when you want to determine (boolean)
> 
> * input passivity (`nu != 0, rho = 0`),
> * output passivity (`nu = 0, rho != 0`)
> 
> now? One would have to use the `get_*_index()` functions. 
> 
> Sorry if I created confusion earlier. I meant to make the function return values intuitive, not to remove functionality.

I believe I've captured this use case now.

Reply:
> No that is fine. This can be defered to later PRs. Just raise a NotImplemented Exception instead of stating "This input is invalid". **Same for I/O passivity index, I guess.**

I'm confused about the bolded part?

Reply:
Sorry, forget the bolded part. Sloppy editing by me. When there is no `get_passive_index()` catchall function you don't need  that.

But *(optional)* nice to have as a future reminder to implement it:

```python
def get_relative_index(sys):
    """Returns the relative passivity index for the system.

    (not implemented yet)
    """
    raise NotImplemented("Relative passivity index not implemented")


def get_combined_io_index(sys):
    """Returns the combined I/O passivity index for the system.

    (not implemented yet)
    """
    raise NotImplemented("Combined I/O passivity index not implemented")


def get_directional_index(sys):
    """Returns the directional passivity index for the system.

    (not implemented yet)
    """
    raise NotImplemented("Directional passivity index not implemented")

```

Reply:
Awesome, I'll take that change.